core/consensus/grandpa/impl/voting_round_impl.hpp

```diff
@@ -0,0 +1,129 @@
+/**
+ * Copyright Soramitsu Co., Ltd. All Rights Reserved.
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#ifndef KAGOME_CORE_CONSENSUS_GRANDPA_IMPL_VOTING_ROUND_IMPL_HPP
+#define KAGOME_CORE_CONSENSUS_GRANDPA_IMPL_VOTING_ROUND_IMPL_HPP
+
+#include "consensus/grandpa/voting_round.hpp"
+
+#include <boost/asio/basic_waitable_timer.hpp>
+#include <boost/signals2.hpp>
+#include "common/logger.hpp"
+#include "consensus/grandpa/environment.hpp"
+#include "consensus/grandpa/grandpa_config.hpp"
+#include "consensus/grandpa/vote_crypto_provider.hpp"
+#include "consensus/grandpa/vote_graph.hpp"
+#include "consensus/grandpa/vote_tracker.hpp"
+
+namespace kagome::consensus::grandpa {
+
+  class VotingRoundImpl : public VotingRound {
+    using OnCompleted = boost::signals2::signal<void(const CompletedRound &)>;
+    using OnCompletedSlotType = OnCompleted::slot_type;
+
+   public:
+    ~VotingRoundImpl() override = default;
+
+    VotingRoundImpl(const GrandpaConfig &config,
+                    std::shared_ptr<Environment> env,
+                    std::shared_ptr<VoteCryptoProvider> vote_crypto_provider,
+                    std::shared_ptr<VoteTracker<Prevote>> prevotes,
+                    std::shared_ptr<VoteTracker<Precommit>> precommits,
+                    std::shared_ptr<VoteGraph> graph,
+                    std::shared_ptr<Clock> clock,
+                    std::shared_ptr<boost::asio::io_context> io_context);
+
+    void onFin(const Fin &f) override;
+
+    void onPrimaryPropose(const SignedPrimaryPropose &primary_propose) override;
+
+    void onPrevote(const SignedPrevote &prevore) override;
+
+    void onPrecommit(const SignedPrecommit &precommit) override;
+
+    bool tryFinalize() override;
+
+    RoundNumber roundNumber() const override;
+
+    void primaryPropose(const RoundState &last_round_state) override;
+
+    void prevote(const RoundState &last_round_state) override;
+
+    void precommit(const RoundState &last_round_state) override;
+
+    inline const RoundState &getCurrentState() const {
+      return cur_round_state_;
+    }
+
+    bool completable() const;
+
+   private:
```

Comment:
Would be nice to provide short comments for private methods.

Reply:
added